geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoVisibilityRowEncoder.scala

```diff
@@ -68,22 +69,31 @@ class KryoVisibilityRowEncoder extends RowEncodingIterator {
     * @return
     */
   private def encodeV3(keys: java.util.List[Key], values: java.util.List[Value]): Value = {
-    // 1 byte for version + 2 bytes for count + 1 byte for size + 4 bytes per attribute for offsets
-    // + 4 bytes for user data offset + null bit set bytes
-    var length = (count + 2 + IntBitSet.size(count)) * 4
 
-    var valueCursor = length // tracks our output position for copying bytes
+    // Calculate length of serialised attributes, excluding attribute values themselves
+    var length = 1 +            // version
+      2 +                       // attribute count
+      1 +                       // offset size
+      count * 4 +               // attribute offsets (will use 4 bytes each) <--- offsets relative to first byte here
+      1 * 4 +                   // user data offset (will use 4 bytes)
+      IntBitSet.size(count) * 4 // null bit set, written in units of ints
+
+    // Tracks our output position for copying attribute value bytes
+    // Begins immediately after the above
+    var valueCursor = length
 
     var i = 0
     while (i < keys.size) {
       val bytes = values.get(i).get
       val input = KryoFeatureDeserialization.getInput(bytes, 1, bytes.length - 1) // skip the version byte
       val metadata = Metadata(input) // read count, size, etc
+      // Column qualifiers tell us which attributes are intended to be populated given visibilities
       keys.get(i).getColumnQualifier.getBytes.foreach { unsigned =>
         val index = java.lang.Byte.toUnsignedInt(unsigned)
         val pos = metadata.setPosition(index)
         val len = metadata.setPosition(index + 1) - pos
-        attributes(index) = (bytes, 4 + pos, len) // the offset is always 4 since we're copying into a new byte array
+        attributes(index) = (bytes, 4 + pos, len) // pos is relative to first byte of attribute offsets, see length calc
+        assert((metadata.nulls.contains(index) && len==0) || (!metadata.nulls.contains(index) && len!=0), s"Metadata null bitset (${metadata.nulls.toString}) and value length ($len) are inconsistent")
```

Comment:
When would this assert fail?  Is this going to stop writers?

Said differently, is anyone going to hit this?

Reply:
asserts usually get compiled out. if you run a unit test through intellij, it will add the `-ea` flag to the compile option to enable assertions. i'm not sure if maven does something similar but probably not since it would have to re-build to remove them.

Reply:
ah, i guess it's a runtime check to enable assertions, not compile time.

Reply:
I'm not sure, that's why the assert is there. It wasn't clear to me from the Kryo encoding docs that a null value was indicated by both the corresponding null bit being set and the length of the attribute being zero (seems redundant?).

I chose assert (Vs require) so they could be disabled at runtime (this might be the default).

I don't mind losing the line tho.

Reply:
hmm, actually i think the length could be zero without it being null, i.e. for an empty string. so yeah, it's probably better to remove the check, or at least the second clause. null implies zero length, but zero length does not imply null.

Reply:
so, looking again, it does seem like any non-null value will write at least one byte, so this will work. however, i realized that we weren't checking for nulls in the initial loop, which would also solve the problem and would probably be more "correct" in that it doesn't rely on an implementation detail of how the bytes are serialized. I.e. `val nulls = IntBitSet(count)` should be moved up above this code block, and then this foreach should be:

```
        val index = java.lang.Byte.toUnsignedInt(unsigned)
        if (metadata.nulls.contains(index)) {
          nulls.add(index)
        } else {
          val pos = metadata.setPosition(index)
          val len = metadata.setPosition(index + 1) - pos
          attributes(index) = (bytes, 4 + pos, len) // the offset is always 4 since we're copying into a new byte array
          length += len
        }
```

then the check down below for zero length doesn't need to be done.

Reply:
actually we don't even need to set the nulls in this loop, since it's set below. just need the `if (!metadata.nulls.contains(index))` guard around setting `attributes(index) = ...`.

Reply:
Yeah, that sounds a much better solution - will make change. I forget why I wasn't looking at the original null bitsets (there was a reason)